dockerworker/renameimage.go

```diff
@@ -0,0 +1,155 @@
+			data, err := ioutil.ReadAll(r)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed to read the metadata file from docker image tar-stream")
+			}
+			switch hdr.Name {
+			case "repositories":
+				data, err = rewriteRepositories(imageName, data)
+			case "manifest.json":
+				data, err = rewriteManifest(imageName, data)
+			default:
+				panic(errors.New("unreachable"))
+			}
+			if err != nil {
+				return nil, nil, err
+			}
+			hdr.Size = int64(len(data))
+			return hdr, bytes.NewReader(data), nil
+		default:
+			if !isAllowedDockerLayerFile(hdr.Name) {
+				return nil, nil, fmt.Errorf(
+					"docker image tar-ball contains illegal file: '%s', consider using an older version of docker", hdr.Name,
+				)
+			}
+			return hdr, r, nil
+		}
+	})
+	// If there was no error and name wasn't written at some point, then we have problem
+	if rwerr == nil && !nameWritten {
+		return fmt.Errorf(
+			"docker image tar-ball did contain 'manifest.json' or 'repositories', docker image is invalid",
+		)
+	}
+	return rwerr
+}
+
+// rewriteRepositories will rewrite the 'repositories' file from a docker image
+// tar-stream such that the image has a single tag named <imageName>:latest
+func rewriteRepositories(imageName string, data []byte) ([]byte, error) {
+	// Rewrite the 'repositories' by reading it as JSON from the input stream
+	var repositories map[string]map[string]string
+	err := json.Unmarshal(data, &repositories)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"the 'repositories' file in the docker image tar-ball is not valid JSON, error: %s",
+			err.Error(),
+		)
+	}
+	// Find the layerID
+	var layerID string
+	for _, tags := range repositories {
+		for _, ID := range tags {
+			// if there are multiple tags pointing to different IDs that's problematic
+			if layerID == "" {
+				layerID = ID
+			} else if layerID != ID {
+				return nil, fmt.Errorf(
+					"the 'repositories' file in the docker image tar-ball contains multiple tags with different layer IDs",
+				)
+			}
+		}
+	}
+	// If there is no imageID we have a problem
+	if layerID == "" {
+		return nil, fmt.Errorf(
+			"the 'repositories' file in the docker image tar-ball did not contain any layer identifiers, image is invalid",
+		)
+	}
+	// Create result of the rewrite and return it
+	data, err = json.Marshal(map[string]map[string]string{
+		imageName: {"latest": layerID}, // map from imageName to layerID
+	})
+	if err != nil {
+		panic(errors.Wrap(err, "json.Marshal failed on map[string]map[string]string that can't happen"))
+	}
+	return data, nil
+}
+
+// Type to read from manifest.json
+type manifestEntry struct {
+	Config   string   `json:"Config,omitempty"`
+	RepoTags []string `json:"RepoTags"`
+	Layers   []string `json:"Layers"`
+}
+
+// rewriteManifest will rewrite the 'manifest.json' file from a docker image
+// tar-stream such that the image has a single tag named <imageName>:latest
+func rewriteManifest(imageName string, data []byte) ([]byte, error) {
```

Comment:
can you explain what the rewrite functions are used for?

Reply:
It is used to rewriter the manifest.json inside the downloaded docker image